File: m/rls/rlb/stwld.c
/* file -  stwld - wildcard compare */
#include "m:\rid\rider.h"
#include "m:\rid\stdef.h"
#include "m:\rid\chdef.h"
/* code -  st_wld - find wildcard substring */
int st_wld(
char *mod ,
char *str )
{ for(;;)  {
    if( !*mod && !*str)return 1;
    if ( *mod == '*') {
      ++mod;
      if( !*mod)return 1;
      while ( *str) {
        if( st_wld (mod,str++))return 1; }
      return 0; }
    if (( *mod == '?')
    ||(*mod == '%')) {
      if( !*str)return 0;
       ++mod, ++str;continue; }
    if( (ch_low (*mod) != ch_low (*str)) != 0)return 0;
    ++mod, ++str;
  } 
} 

File: m/rls/stwld.r
file	stwld - wildcard compare
include	rid:rider
include	rid:stdef
include	rid:chdef

code	st_wld - match wildcard string

;	*	Wild string
;	% or ?	Wild character

  func	st_wld
	mod : * char~			; model substring with *? to find
	str : * char~			; string to search
	()  : int			; matched
  is  repeat				; big loop
	fine if !*mod && !*str		; all over
	if *mod eq '*'			; got a wildcard
	   ++mod			;
	   fine if !*mod		; ...* matchs anything
	   while *str			; attempt the remainder
	   .. fine if st_wld (mod,str++);
	.. fail				;
	if *mod eq '%'			;
	|| *mod eq '?'			;
	   fail if !*str		; nothing to match
	.. next ++mod, ++str		; skip single
	ch_low (*mod) ne ch_low (*str)	; compare one
	fail if ne			; (ne => true)
	++mod, ++str			; next pair
      forever
  end
